File: py/packages/genkit-vertexai/src/genkit_vertexai/model_garden/anthropic.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0


"""Anthropic models."""

from collections.abc import Awaitable, Callable
from typing import cast

from anthropic import AsyncAnthropic, AsyncAnthropicVertex
from genkit_anthropic.config import AnthropicConfig
from genkit_anthropic.models import AnthropicModel
from pydantic import ConfigDict
from pydantic.config import JsonDict

from genkit import ModelInfo, ModelRequest, ModelResponse, Supports
from genkit.plugin_api import ActionRunContext, ModelConfig, loop_local_client


def _vertex_anthropic_config_schema_extra(schema: JsonDict) -> None:
    """Drop options Vertex Model Garden cannot honor from the advertised schema."""
    base_extra = AnthropicConfig.model_config.get('json_schema_extra')
    if callable(base_extra):
        cast(Callable[[JsonDict], None], base_extra)(schema)
    properties = schema.get('properties')
    if isinstance(properties, dict):
        properties.pop('apiKey', None)


class VertexAnthropicConfig(AnthropicConfig):
    """Anthropic config for Vertex Model Garden.

    ``apiKey`` is omitted because :class:`AsyncAnthropicVertex` authenticates
    with ambient Google credentials and ignores a per-request Anthropic key.
    """

    model_config = ConfigDict(**{
        **AnthropicConfig.model_config,
        'json_schema_extra': _vertex_anthropic_config_schema_extra,
    })


class AnthropicModelGarden:
    """Manages integration with Anthropic models on Vertex AI Model Garden."""

    def __init__(
        self,
        model: str,
        location: str,
        project_id: str,
    ) -> None:
        """Initializes the AnthropicModelGarden instance.

        Args:
            model: The name of the specific model to be used from Model Garden
                in the way <publisher>/<model> (e.g., 'anthropic/claude-3-5-sonnet-v2@20241022').
            location: The Google Cloud region where the Model Garden service
                is hosted (e.g., 'us-central1').
            project_id: The Google Cloud project ID where the Model Garden
                model is deployed.
        """
        self.name = model
        self._runtime_client = loop_local_client(lambda: AsyncAnthropicVertex(region=location, project_id=project_id))
        # Strip 'anthropic/' prefix for the model passed to Anthropic SDK
        clean_model_name = model.removeprefix('anthropic/')
        self._model_name = clean_model_name

    def get_handler(self) -> Callable[[ModelRequest, ActionRunContext], Awaitable[ModelResponse]]:
        """Returns the generate handler function for this model.

        Returns:
            The handler function that can be used as an Action's fn parameter.
        """

        async def _generate(request: ModelRequest, ctx: ActionRunContext) -> ModelResponse:
            model = AnthropicModel(
                model_name=self._model_name,
                client=cast(AsyncAnthropic, self._runtime_client()),
            )
            return await model.generate(request, ctx)

        return _generate

    def get_model_info(self) -> ModelInfo:
        """Returns the model information/metadata for this model.

        Returns:
            ModelInfo with the model's capabilities.
        """
        return ModelInfo(
            label=f'ModelGarden - {self.name}',
            supports=Supports(
                multiturn=True,
                media=True,
                tools=True,
                system_role=True,
                output=['text', 'json'],
            ),
        )

    @staticmethod
    def get_config_schema() -> type[ModelConfig]:
        """Returns the config schema for this model type."""
        return VertexAnthropicConfig

File: go/plugins/vertexai/modelgarden/anthropic.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

package modelgarden

import (
	"context"
	"fmt"
	"sync"

	"github.com/anthropics/anthropic-sdk-go"
	"github.com/anthropics/anthropic-sdk-go/vertex"
	"github.com/firebase/genkit/go/ai"
	"github.com/firebase/genkit/go/core/api"
	"github.com/firebase/genkit/go/genkit"

	"github.com/firebase/genkit/go/plugins/internal"
	ant "github.com/firebase/genkit/go/plugins/internal/anthropic"
)

const pluginName = "vertex-model-garden-anthropic"

// Anthropic is a Genkit plugin for interacting with Anthropic models in Vertex AI Model Garden
type Anthropic struct {
	ProjectID string // Google Cloud project to use for Vertex AI. If empty, the value of the environment variable GOOGLE_CLOUD_PROJECT or GCLOUD_PROJECT will be consulted in that order
	Location  string // Location of the Vertex AI service. If empty, the value of the environment variable GOOGLE_CLOUD_LOCATION or GOOGLE_CLOUD_REGION will be consulted in that order

	client  anthropic.Client // Client for the model garden service
	mu      sync.Mutex       // Mutex to control access
	initted bool             // Whether the plugin has been initialized
}

// Name returns the name of the plugin
func (a *Anthropic) Name() string {
	return pluginName
}

// Init initializes the VertexAI Model Garden for Anthropic plugin and all its known models.
// After calling Init, you may call [DefineModel] to create and register any additional models.
func (a *Anthropic) Init(ctx context.Context) []api.Action {
	a.mu.Lock()
	defer a.mu.Unlock()
	if a.initted {
		panic("plugin already initialized")
	}

	projectID, location := resolveVertexMaasEnv(a.ProjectID, a.Location)

	c := anthropic.NewClient(
		vertex.WithGoogleAuth(ctx, location, projectID, "https://www.googleapis.com/auth/cloud-platform"),
	)
	a.client = c
	a.initted = true

	// Claude models in VertexAI cannot be listed using the Anthropic SDK
	// Models must be defined manually
	var actions []api.Action
	for name, opts := range AnthropicModels {
		// The catalog stores bare display names ("Claude Opus 4.6"). Prefix
		// the provider so these sit alongside the other Vertex AI models in a
		// picker instead of looking like they came from somewhere else.
		opts.Label = internal.ProviderLabel(ant.DisplayName(provider), opts.Label)
		actions = append(actions, ant.NewModel(a.client, provider, name, opts))
	}

	return actions
}

// AnthropicModel returns the [ai.Model] with the given id.
// It returns nil if the model was not defined
func AnthropicModel(g *genkit.Genkit, id string) ai.Model {
	return genkit.LookupModel(g, api.NewName(provider, id))
}

// DefineModel builds a Model Garden Claude model and returns it. It does not
// register the model: generation resolves a model from its name, so passing
// the result to ai.WithModel contributes only that name and serves the request
// with a model resolved from it instead. Pass the result to
// [genkit.RegisterAction] to make these capabilities the ones used.
func (a *Anthropic) DefineModel(name string, opts *ai.ModelOptions) (ai.Model, error) {
	a.mu.Lock()
	defer a.mu.Unlock()
	if !a.initted {
		return nil, fmt.Errorf("modelgarden anthropic: plugin not initialized")
	}
	if opts == nil {
		return nil, fmt.Errorf("DefineModel called with nil ai.ModelOptions")
	}
	return ant.NewModel(a.client, provider, name, *opts), nil
}
